README
COMPILING
 - Compiling instructions can be found in the INSTALL file

USAGE
	start an ecosim
	> ./src/critterding

	start a race
	> ./src/critterding --profile profiles/race

TERMINOLOGY

 - Adam
      : new (random) critter

 - Adam Distance (ad)
      : generations with mutations since adam

OPTIONS

  option                                                    [default] [range]      [comment]

  --autoload                                                  0         0-1          autoload critters from ~/.critterding/load
  --autoloadlastsaved                                         0         0-1          autoload critters from ~/.critterding/lastsaved
  --benchmark                                                 0         0-1          run the critterding benchmark
  --benchmark_frames                                          1000      1-2147483647 length of benchmark in frames
  --body_constraintpositionrange                              60        0-100        amount muscles can deviate from the center of a plane (%)
  --body_costhavingbodypart                                   1000      0-2147483647 cost of having a bodypart (1/100000 energy)
  --body_costtotalweight                                      1000      0-2147483647 cost multiplier of total body weight (1/100000 energy)
  --body_maxbodyparts                                         30        0-2147483647 maximum body parts per critter
  --body_maxbodypartsatbuildtime                              8         1-2147483647 maximum body parts for a new critter
  --body_maxbodypartsize                                      100       1-2147483647 maximum size of a critters body part
  --body_maxconstraintangle                                   31415     0-2147483647 maximum amount of rotation on a contraint
  --body_maxconstraintlimit                                   14000     0-2147483647 maximum limit of a contraint
  --body_maxheadsize                                          50        1-2147483647 maximum size of a critters head
  --body_maxmutations                                         4         1-2147483647 maximum mutations on a body mutant
  --body_minbodyparts                                         2         0-2147483647 minimum body parts per critter
  --body_minbodypartsatbuildtime                              4         1-2147483647 minimum body parts for a new critter
  --body_minbodypartsize                                      20        1-2147483647 minimum size of a critters body part
  --body_minheadsize                                          15        1-2147483647 minimum size of a critters head
  --body_mutationrate                                         160       0-1000       rate of body mutation (permille)
  --body_percentmutateeffectaddbodypart                       2         0-100        chance of adding a body part
  --body_percentmutateeffectchangecolor                       2         0-100        chance of changing body color
  --body_percentmutateeffectchangecolor_slightly              5         0-100        chance of slightly changing body color
  --body_percentmutateeffectchangeconstraintangles            2         0-100        chance of changing a joints position angles
  --body_percentmutateeffectchangeconstraintangles_slightly   5         0-100        chance of changing a joints position angles
  --body_percentmutateeffectchangeconstraintlimits            2         0-100        chance of changing a joints motion limits
  --body_percentmutateeffectchangeconstraintlimits_slightly   5         0-100        chance of slightly changing a joints motion limits
  --body_percentmutateeffectchangeconstraintposition          2         0-100        chance of changing a joints position
  --body_percentmutateeffectchangeconstraintposition_slightly 0         0-100        chance of slightly changing a joints position
  --body_percentmutateeffectremovebodypart                    2         0-100        chance of removing a body part
  --body_percentmutateeffectrepositionhead                    5         0-100        chance of repositioning head
  --body_percentmutateeffectresizebodypart                    2         0-100        chance of resizing a body part
  --body_percentmutateeffectresizebodypart_slightly           5         0-100        chance of slightly resizing a body part
  --body_percentmutateeffectresizehead                        2         0-100        chance of resizing a head
  --body_percentmutateeffectresizehead_slightly               5         0-100        chance of slightly resizing a head
  --body_selfcollisions                                       0         0-1          critters can exploit physics glitches
  --brain_costfiringmotorneuron                               1000      0-2147483647 cost of firing a motor neuron
  --brain_costfiringneuron                                    50        0-2147483647 cost of firing a neuron
  --brain_costhavingneuron                                    100       0-2147483647 cost of having a neuron (1/100000 energy)
  --brain_costhavingsynapse                                   20        0-2147483647 cost of having a synapse
  --brain_maxfiringthreshold                                  12        1-2147483647 maximum firingthreshold of a neuron
  --brain_maxmutations                                        10        1-2147483647 maximum mutations on a brain mutant
  --brain_maxneurons                                          1000      1-2147483647 maximum neurons per critter
  --brain_maxneuronsatbuildtime                               150       1-2147483647 maximum neurons for a new critter
  --brain_maxplasticitystrengthen                             1000      1-2147483647 maximum weight by which plastic synapses strengthen
  --brain_maxplasticityweaken                                 10000     1-2147483647 maximum weight by which plastic synapses weaken
  --brain_maxsynapses                                         1000      1-2147483647 maximum synapses per neuron
  --brain_maxsynapsesatbuildtime                              200       1-2147483647 maximum synapses for a new neuron of a new critter
  --brain_minfiringthreshold                                  2         1-2147483647 minimum firingthreshold of a neuron
  --brain_minneuronsatbuildtime                               10        1-2147483647 minimum neurons for a new critter
  --brain_minplasticitystrengthen                             100       1-2147483647 minimum weight by which plastic synapses strengthen
  --brain_minplasticityweaken                                 1000      1-2147483647 minimum weight by which plastic synapses weaken
  --brain_minsynapses                                         1         1-2147483647 minimum synapses per neuron
  --brain_minsynapsesatbuildtime                              10        1-2147483647 minimum synapses for a new neuron
  --brain_mutate_maxfiringthreshold                           0         0-1          mutate this value
  --brain_mutate_minfiringthreshold                           0         0-1          mutate this value
  --brain_mutate_mutateeffects                                0         0-1          mutate mutation effects
  --brain_mutate_percentchanceconsistentsynapses              0         0-1          mutate this value
  --brain_mutate_percentchanceinhibitoryneuron                0         0-1          mutate this value
  --brain_mutate_percentchanceinhibitorysynapses              0         0-1          mutate this value
  --brain_mutate_percentchancemotorneuron                     0         0-1          mutate this value
  --brain_mutate_percentchanceplasticneuron                   0         0-1          mutate this value
  --brain_mutate_percentchancesensorysynapse                  0         0-1          mutate this value
  --brain_mutate_plasticityfactors                            0         0-1          mutate min/max plasticity values
  --brain_mutationrate                                        160       0-1000       rate of brain mutation (permille)
  --brain_percentchanceconsistentsynapses                     0         0-100        percent chance a neurons synapses are all inhibitory or excitatory
  --brain_percentchanceinhibitoryneuron                       50        0-100        percent chance a neuron is inhibotory
  --brain_percentchanceinhibitorysynapses                     50        0-100        percent chance a synapse is inhibitory
  --brain_percentchancemotorneuron                            100       0-100        percent chance a neuron is a motor neuron
  --brain_percentchanceplasticneuron                          20        0-100        percent chance a neuron has plastic synapses
  --brain_percentchancesensorysynapse                         20        0-100        percent change a synapse connects to an input
  --brain_percentmutateeffectaddneuron                        2         0-100        chance of adding a neuron
  --brain_percentmutateeffectaddsynapse                       10        0-100        chance of adding a synapse
  --brain_percentmutateeffectaltermutable                     1         0-100        mutate value of a mutatable option
  --brain_percentmutateeffectalterneuron                      5         0-100        chance of altering a neuron
  --brain_percentmutateeffectremoveneuron                     2         0-100        chance of removing a neuron
  --brain_percentmutateeffectremovesynapse                    10        0-100        chance of removing a synapse
  --camera_mode                                               0         0-3          camera mode
  --camera_smoothfactor                                       860       0-1000       camera smoothing factor (permille)
  --camerasensitivity_look                                    950       1-10000      sensitivity of the camera when looking around
  --camerasensitivity_move                                    600       1-10000      sensitivity of the camera when moving
  --colormode                                                 0         0-1          colors genetically exact critters identically
  --critter_autoexchangeinterval                              0         0-2147483647 save critters every n seconds
  --critter_autosaveinterval                                  0         0-2147483647 save critters every n seconds
  --critter_enableomnivores                                   0         0-1          enables critters to eat each other
  --critter_insertevery                                       0         0-2147483647 inserts a random critter every n frames
  --critter_maxenergy                                         5000      1-2147483647 maximum amount of energy a critter has
  --critter_maxlifetime                                       32000     1-2147483647 maximum number of frames a critter lives
  --critter_minenergyproc                                     3000      1-2147483647 energy a critters needs to procreate
  --critter_mutate_maxlifetime                                0         0-2147483647 chance of mutating the maximum number of frames a critter lives
  --critter_mutate_minenergyproc                              0         0-2147483647 chance of mutating the energy a critters needs to procreate
  --critter_procinterval                                      100       0-2147483647 minimum frames between procreations
  --critter_retinasize                                        16        1-50         size of a critters eye retina
  --critter_sightrange                                        500       1-2147483647 distance a critter can see (10 = 1 worldsize)
  --critter_startenergy                                       3000      1-2147483647 energy a new critter (adam) starts with
  --drawdebug                                                 0         0-5          debug drawing mode
  --drawscene                                                 1         0-1          draw the scene
  --energy                                                    1200      0-2147483647 energy in the system by number of food cubes
  --exit_if_empty                                             0         0-1          exit simulation if there are no critters
  --food_maxenergy                                            1500      1-2147483647 maximum amount of energy a food unit has
  --food_maxlifetime                                          32000     1-2147483647 maximum number of frames a food unit exists
  --fpslimit                                                  0         0-1          enable fps limiter
  --fpslimit_frames                                           60        1-1000       frames per second for the fps limiter
  --fsX                                                       1920      1-2147483647 fullscreen resolution X
  --fsY                                                       1200      1-2147483647 fullscreen resolution Y
  --glsl                                                      1         0-1          enable GLSL
  --hdr                                                       0         0-1          enable HDR (experimental)
  --light_ambientB                                            8         0-255        lighting setting
  --light_ambientG                                            8         0-255        lighting setting
  --light_ambientR                                            8         0-255        lighting setting
  --light_ambient_modelB                                      8         0-255        lighting setting
  --light_ambient_modelG                                      8         0-255        lighting setting
  --light_ambient_modelR                                      8         0-255        lighting setting
  --light_attenuation_constant                                200       0-2147483647 lighting setting
  --light_attenuation_linear                                  200       0-2147483647 lighting setting
  --light_attenuation_quadratic                               200       0-2147483647 lighting setting
  --light_diffuseB                                            255       0-255        lighting setting
  --light_diffuseG                                            255       0-255        lighting setting
  --light_diffuseR                                            255       0-255        lighting setting
  --light_specularB                                           255       0-255        lighting setting
  --light_specularG                                           255       0-255        lighting setting
  --light_specularR                                           255       0-255        lighting setting
  --map                                                       easy                   map to load (directory)
  --mincritters                                               20        0-1000       minimum number of critters
  --population_double                                         0         0-2147483647 duplicate all critters if population reaches n
  --population_eliminate_portion                              160       2-2147483647 eliminate a percentage of critters if population reaches n
  --population_eliminate_portion_bodymutationratechange       0         0-1          change mutationrate of brain between min and max
  --population_eliminate_portion_bodymutationratemax          0         0-100        maximum mutationrate of brain
  --population_eliminate_portion_bodymutationratemin          0         0-100        minimum mutationrate of brain
  --population_eliminate_portion_brainmutationratechange      0         0-1          change mutationrate of brain between min and max
  --population_eliminate_portion_brainmutationratemax         0         0-100        maximum mutationrate of body
  --population_eliminate_portion_brainmutationratemin         0         0-100        minimum mutationrate of body
  --population_eliminate_portion_decrenergy                   2         0-100        decrease energy by n when killhalfat triggers
  --population_eliminate_portion_decrmaxlifetimepct           0         0-100        decrease critter_maxlifetime by n when killhalfat triggers
  --population_eliminate_portion_incrworldsizeX               0         0-100        increase worldsizeX by n when killhalfat triggers
  --population_eliminate_portion_incrworldsizeY               0         0-100        increase worldsizeY by n when killhalfat triggers
  --population_eliminate_portion_incrworldsizeZ               0         0-100        increase worldsizeZ by n when killhalfat triggers
  --population_eliminate_portion_percent                      35        1-100        eliminate n% of critters if population reaches population_eliminate_portion
  --population_limit_energy                                   150       0-2147483647 population limit
  --population_limit_energy_percent                           100       0-100        if population limit is exceeded drop energy to percentage
  --race                                                      0         0-1          enable race simulation
  --roundworld                                                0         0-1          enable round planet
  --skybox                                                    night-day/skydome3.obj skybox to load (obj file)
  --skybox_rotationX                                          0         0-2147483647 rotational speed of skybox along the X axis
  --skybox_rotationY                                          0         0-2147483647 rotational speed of skybox along the Y axis
  --skybox_rotationZ                                          0         0-2147483647 rotational speed of skybox along the Z axis
  --skybox_scale                                              200       0-2147483647 scale of the skybox (100=1)
  --startseed                                                 0         0-2147483647 startseed for random number generation
  --threads                                                   1         1-2147483647 threads to use
  --window_color_bits                                         24        16-32        window framebuffer color bits per pixel
  --window_depth_bits                                         24        8-32         window framebuffer depth bits per pixel
  --window_doublebuffer                                       1         0-1          double buffered window
  --window_fullscreen                                         0         0-1          enable fullscreen window
  --window_hardware_gl                                        1         0-1          enable hardware GL window
  --window_height                                             600       1-2147483647 window client height in pixels
  --window_multisamples                                       0         0-16         window framebuffer multisamples
  --window_resizable                                          1         0-1          enable resizable window
  --window_stencil_bits                                       0         0-32         window framebuffer stencil bits per pixel
  --window_vsync                                              0         0-1          vsynced window
  --window_width                                              900       1-2147483647 window client width in pixels
  --worldsizeX                                                200       1-2147483647 size of the world along axis X
  --worldsizeY                                                100       1-2147483647 size of the world along axis Y
  --worldsizeZ                                                120       0-2147483647 size of the world along axis Z

 To save the default settings to a profile, press "s" in the simulation.  
 It will be saved to "profiles/[timestamp]" and can be loaded by using "--profile profiles/[timestamp]"

 Press F1 in the simulation for more information about keys.



Bob Winckelmans <bob.winckelmans@telenet.be>
Sam Porter
IRC:
   channel: #critterding
   server: irc.freenode.org

Changelog
1.0-beta14
------------
custom maps (wavefront)
normal mapping
textures
customizable lightsource
critter cameras
select species from species panel
weight calculation fixes
debug drawer for physics

1.0-beta13.1
------------
make install
bzip compression on critters
OpenMP is now optional
removed libusb dependency
autoexchange segfault fix
boost linking fix
a few cleanups

1.0-beta13
----------
bodypart selfcollisions fix (thanks ScribbleJ)
roundworld gravity fix (thanks Sam Porter)
refactored critter brains, muscles
refactored brainview & species windows
upgrade bullet physics and link it statically
decrease energy by the unit instead of percent
added energy cost of bodyparts and weight
mutationrates are now permille instead of percent
opengllighting fix & more contrasting critter colors
food count reporting fix
critters can see if critter is of the same species or not
customisable keybindings ( settings.xml )
threading fixes
switched to cmake
fullscreen & window resizing fixes
hundreds of fixes and cleanups


1.0-beta12.1
------------
catch segfault and warn when using headless without critter_raycastvision
some fixes and cleanups

1.0-beta12
----------
critter selection and actions (kill, duplicate)
new panels: neural net brainviewer, hud, species list
touchingcritter brain input fix
profile saves to ~/critterding/save/(profile)/(profile).pro
toggle rendering of gui and scene (h & r)
font switch to DejaVuSans
an icon thanks to jrabbit
new options:
 --roundworld: a round planet
 --benchmark: times a scene of 10000 frames
 --headless: console mode
 --startseed: seed for the random number generator
 --killhalf_incrworldsizeX/Y: option to increase worldsize when killhalf triggers
 --killhalf_decrmaxlifetimepct: option to decrease critter maxlifetime when killhalf triggers
 --critter_raycastvision: raycast vision for critters
 --threads, number of threads, making openmp (comes with gcc4.2+) a requirement
build system: 
 if available, build against system ftgl (pass --disable-system-ftgl to use internal)
 make install
many fixes, cleanups & changes

1.0-beta11
---------
improved keyboard / mouse input handling
improved gui with panels for realtime setting manipulation,
critters vs food graph,
exit panel (esc no longer exits)
a few more body mutations
fix: critter autosaving timer
fix: crashes on exit
many more fixes and cleanups

1.0-beta10
---------
changed the physics backend to Bullet Physics
changed input and window backend to SDL
a racing mode
mouse support for looking and interactivity
profile saving (patch from Eric Burton)
fullscreen support
many improvements, fixes and cleanups overall

1.0-beta9
---------
dedicated skincolors for herbivores(Red/Blue) / carnivores(Green/Blue)
bullets no longer shoot through walls
world is no longer fixed to a square
fps limiter
introduced text rendering with fglx
improved synaptic plasticity
slightly improved wall system
--autoload option
cleaned up critter files
color inputneurons get less discrete input
many fixes and cleanups

1.0-beta8
---------
new: custom settings profiles, critter autosaving, exit-if-empty
corpses and carrying are disabled at default
various fixes and cleanups

1.0-beta7
---------
critters are now classified as herbivore or carnivore
corpse is poison to a herbivore/green food is poison to a carnivore
fixed a alpha color bug in critter vision (vidia cards)
engine improvements: backface & frustrum culling + display lists
more efficient copying of brains at procreation time
simplified (more usable) camera with adjustable sensitivity
critters can now pick up and drop food

1.0-beta6
---------
age input neurons
started testing a corpse entity
discovered and fixed retina to input mapping bug
visibility enhancement by allowing only bright colors
bullet hit: half of lifetime lost
camera autopositioning according to world size
--small, --medium, --big and --huge options

1.0-beta5
---------
Options have boundary checks now
improved command line option handling
f1 pauses and prints settings / keys
increased evolution rate by tuned defaults settings
fixed README mistakes

1.0-beta4
---------
adaptable colorneurons, default is now 3 instead of 2
adaptable retinasize, default is nog 7 instead of 10
all (38) initial brain settings have been made adaptable
provided an option to flip newborn critters 180 degrees
simple --help implemented
many minor fixes and cleanups

1.0-beta3
---------
fixes minor memleaks & speedups
brain.clearInputs() implementation                             
critter & food size have been made adaptable
make sure we only load files with extension cr
Retinas Per Row is now adaptable
default critter sight distance reduced from 50 to 40: performance boost
follow camera handles critter sight distance
cleaned up settings reporting
updated README
Start of a Changelog


benchmark.sh
# > benchmark result : 
# >   frames      :5000
# >   time        :21.077 seconds
# >   seed        :13
# >   checksum    :1094.2192
# >   average fps :237.3203 fps


# > benchmark result : 
# >   frames      :5000
# >   time        :20.893 seconds
# >   seed        :13
# >   checksum    :1094.2192
# >   average fps :239.41033 fps

# ./src/critterding --glsl 1 --roundworld 1 --worldsizeX 50 --autoload 1 --benchmark 1 --benchmark_frames 5000 --startseed 13 --threads 3 --energy 400 | grep '>'



#############################################################################"
# > benchmark result : 
# >   frames      :3000
# >   time        :12.023 seconds
# >   seed        :13
# >   checksum    :4080.7397
# >   average fps :249.72964 fps

# 	./src/critterding --glsl 1 --roundworld 1 --worldsizeX 30 --benchmark 1 --benchmark_frames 3000 --startseed 13 --threads 2 --energy 1000 --mincritters 4 | grep '>'
#############################################################################"


#############################################################################"
# > benchmark result : 
# >   frames      :8000
# >   time        :59.933 seconds
# >   seed        :13
# >   checksum    :7472.5518
# >   average fps :133.51576 fps

# 	./src/critterding --glsl 1 --roundworld 1 --worldsizeX 40 --benchmark 1 --benchmark_frames 8000 --startseed 13 --threads 2 --energy 1000 --mincritters 40 | grep '>'
#############################################################################"



#############################################################################"
# > benchmark result : 
# >   frames      :3000
# >   time        :12.023 seconds
# >   seed        :13
# >   checksum    :4080.7397
# >   average fps :249.72964 fps

# 	./src/critterding --glsl 1 --roundworld 0 --benchmark 1 --benchmark_frames 2000 --startseed 130 --threads 4 --energy 10 --mincritters 80 --drawscene 0 | grep '>'
	./src/critterding --glsl 1 --roundworld 0 --benchmark 1 --benchmark_frames 2000 --startseed 130 --threads 4 --energy 10 --mincritters 80 --drawscene 0
#############################################################################"

